### ISSUE TITLE
Extensive config validator

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/stryker#438]
Right now: we're reading the config without enforcing all properties have the correct types. We should validate the config to prevent frustrating mistakes and a lot of time loss by consumers.

**Comment by hrb90** [comment]
:wave: I'd be interested in helping with this, if this is still a feature we want and you think it'd be a reasonable first issue to tackle for someone without much open source experience

**Comment by simondel** [comment]
Awesome! Do you have any experience with TypeScript?

This repository is a mono repo. It contains all the packages that the core stryker team actively maintains. You can run commands such as `npm install`, `npm run build` and `npm test` from the root of the repo and the commands will be executed in each package. You can also execute the commands in every package individually to speed up things like testing. You may also want to read our [contributing guide](https://github.com/stryker-mutator/stryker/blob/master/CONTRIBUTING.md).

The config itself is defined in the [stryker-api](https://github.com/stryker-mutator/stryker/blob/master/packages/stryker-api/src/config/Config.ts) and you can find more info on it in the [readme for stryker](https://github.com/stryker-mutator/stryker/tree/master/packages/stryker#configuration). We have a [ConfigValidator](https://github.com/stryker-mutator/stryker/blob/master/packages/stryker/src/ConfigValidator.ts) that is already doing some basic validation.

If you need any help, feel free to let us know!

**Comment by nicojs** [comment]
Awesome! Let's focus this issue only on the main stryker config and ignore any additional config used by plugins. That part is fairly maintained. Indeed, the `ConfigValidator` class is a great place to do it. 

For example: `stryker-typescript` uses the additional settings `tsconfig` and `tsconfigFile`. Let's ignore those for now. Once this issue is resolved we could move to plugin config validation. Once we have all that we can give an error on any additional config we don't support and even give suggestions if someone misspelled a config setting.

**Comment by hrb90** [comment]
Found a little bit of time to look at this this weekend, and I have a few questions.

* The existing private validation methods check the `typeof` a couple of the fields of the config. Do we want to do this for all of them? If so, how would I go about testing the expected behavior, since TypeScript rightly won't compile the code if I mess up the types?

* There are a few places where we're expecting the user to have a certain package installed depending on what's in their config; it seems like we're checking that only when we need it (e.g. stryker looks for the mutator only when the initial test run passes.) Do we want to do those checks when we validate the config?

* There's some validation logic in the `ConfigReader` class; do we want to move that to the `ConfigValidator`, or are we doing the validation in the `ConfigReader` to avoid breaking something in between?

Thanks!

**Comment by nicojs** [comment]
> Do we want to do this for all of them? 

Yes, let's do that, but don't check for additional fields that might possibly be there.

> If so, how would I go about testing the expected behavior, since TypeScript rightly won't compile the code if I mess up the types?

I think the best thing would be to cast it as an any. For example: `const brokenConfig = { files: 'string' } as any;`. There simply is no other way fool typescript into not throwing a compile error. A cool way of doing it might be:

```
function createBrokenConfig(key: keyof StrykerOptions, value: any): any {
   return { [key]: value };
}
```

Using this function ensures you're not making a typo in the property key name itself.

>  it seems like we're checking that only when we need it (e.g. stryker looks for the mutator only when the initial test run passes.). Do we want to do those checks when we validate the config?

I would say yes. In the future, we might add some commands, like `stryker run --dry` or something, which will signal intent of the user more clearly. For now: lets fail fast so the user knows what to do. 

> There's some validation logic in the `ConfigReader` class; do we want to move that to the `ConfigValidator`

Yes, let's move that. The validation in the `ConfigReader` class is some older code that I forgot about when creating the `ConfigValidator`. Good catch!

**Comment by hrb90** [comment]
I have a few tests left to write but otherwise I think I'm done with validating that things are the right type and that the ones that should take one of a finite number of values in fact do so.

Can you tell me where I should look to begin writing validations to check that we have all the plugins/transpilers/etc. that our config file claims we do?

**Comment by nicojs** [comment]
@hrb90 sounds good! 

I think we should open up another issue&PR for validating of plugin configuration. That will keep the PR's easier to manage. I will create a branch later today demonstrating how plugins should be validating.

Basically, every plugin should validate the properties himself and communicate back what properties he expects so that Stryker can give an error for every property that was unexpected.

**Comment by nicojs** [comment]
See #548

**Comment by hrb90** [comment]
Awesome, I'll finish up the stuff from this issue and open a PR tomorrow at latest.

**Comment by hrb90** [comment]
Opened a PR: #549

**Comment by nicojs** [comment]
Closed since PR #549
